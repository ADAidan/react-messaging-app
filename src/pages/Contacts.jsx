/* eslint-disable no-underscore-dangle */
import * as React from "react";
import axios from "axios";
import Container from "@mui/material/Container";
import Stack from "@mui/material/Stack";
import Toolbar from "@mui/material/Toolbar";
import Divider from "@mui/material/Divider";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { setOnlineAmount } from "../features/onlineCounter/onlineCounterSlice";
import { setContactsAmount } from "../features/contactsCounter/contactsCounterSlice";
import { setPendingAmount } from "../features/pendingCounter/pendingCounterSlice";
import socket from "../socket";
import ContactTabs from "../components/ContactTabs";
import SearchBar from "../components/SearchBar";
import ContactCard from "../components/ContactCard";

function Contacts() {
  const userId = sessionStorage.getItem("user");
  const [userContacts, setUserContacts] = React.useState([]);
  const [userPending, setUserPending] = React.useState([]);
  const [filteredContacts, setFilteredContacts] = React.useState([]);
  const [searchValue, setSearchValue] = React.useState("");
  const [searchedContacts, setSearchedContacts] = React.useState([]);
  const [selectedTab, setSelectedTab] = React.useState(0);

  const dispatch = useDispatch();
  const navigate = useNavigate();

  React.useEffect(() => {
    axios
      .get("http://localhost:3000/users/protected", {
        withCredentials: true,
      })
      .then(() => {
        // eslint-disable-next-line no-console
        console.log("accessed protected route");
      })
      .catch(() => {
        navigate("/logout");
      });
  }, []);

  // Sort pending contacts by status and username
  const sortPendingContacts = (pending) => {
    const pendingContacts = pending;
    const incomingRequests = pendingContacts.filter(
      (user) => user.status === "incoming contact request",
    );
    const outgoingRequests = pendingContacts.filter(
      (user) => user.status === "outgoing contact request",
    );

    if (incomingRequests.length) {
      incomingRequests.sort((a, b) => a.username.localeCompare(b.username));
    }
    if (outgoingRequests.length) {
      outgoingRequests.sort((a, b) => a.username.localeCompare(b.username));
    }
    const pendingRequests = incomingRequests.concat(outgoingRequests);
    return pendingRequests;
  };

  // get user contacts
  React.useEffect(() => {
    if (!userId) return;
    socket.emit("JoinRoom", userId);
    const getContactData = async () => {
      try {
        const response = await axios.get(
          `http://localhost:3000/users/${userId}/contacts`,
        );
        const contactData = response.data;
        setUserContacts(contactData);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Axios Error:", error);
      }
    };
    getContactData();
  }, []);

  // Socket event listener for changing user status
  React.useEffect(() => {
    const handleChangeUserStatus = (data) => {
      setUserContacts((prevContacts) => {
        const updatedContacts = prevContacts.map((contact) =>
          contact._id === data.id
            ? { ...contact, status: data.status }
            : contact,
        );
        return updatedContacts;
      });
    };
    socket.on("ChangeUserStatus", handleChangeUserStatus);
    return () => {
      socket.off("ChangeUserStatus", handleChangeUserStatus);
    };
  }, [userContacts]);

  // get user pending contacts
  React.useEffect(() => {
    if (!userId) return;
    const getPendingData = async () => {
      const pendingContacts = [];
      try {
        const response = await axios.get(
          `http://localhost:3000/users/${userId}/pending`,
        );
        const pendingData = response.data;
        for (let i = 0; i < pendingData.length; i += 1) {
          pendingContacts.push(pendingData[i].user);
          pendingContacts[i].status =
            `${pendingData[i].pendingStatus} contact request`;
        }
        const sortedPending = sortPendingContacts(pendingContacts);
        setUserPending(sortedPending);
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Axios Error:", error);
      }
    };
    getPendingData();
  }, []);

  // Socket event listener for deleting contact
  React.useEffect(() => {
    const handleDeleteContact = (data) => {
      setUserContacts((prevContacts) => {
        const updatedContacts = prevContacts.filter(
          (contact) => contact._id !== data._id,
        );
        return updatedContacts;
      });
    };
    socket.on("DeleteContact", handleDeleteContact);
    return () => {
      socket.off("DeleteContact", handleDeleteContact);
    };
  }, [userContacts]);

  // Socket event listener for sending contact request
  React.useEffect(() => {
    const handleUpdatePendingContacts = (data) => {
      setUserPending((prevPending) => {
        const updatedPending = prevPending.concat(data);
        return updatedPending;
      });
    };
    socket.on("UpdatePendingContacts", handleUpdatePendingContacts);
    return () => {
      socket.off("UpdatePendingContacts", handleUpdatePendingContacts);
    };
  }, [userPending]);

  // Socket event listener for accepting contact request
  React.useEffect(() => {
    const handleAcceptPendingContacts = (data) => {
      setUserContacts((prevContacts) => {
        const updatedContacts = prevContacts.concat(data);
        return updatedContacts;
      });
      setUserPending((prevPending) => {
        const updatedPending = prevPending.filter(
          (pending) => pending._id !== data._id,
        );
        return updatedPending;
      });
    };
    socket.on("AcceptContactRequest", handleAcceptPendingContacts);
    return () => {
      socket.off("AcceptContactRequest", handleAcceptPendingContacts);
    };
  }, [userContacts, userPending]);

  // Socket event listener for rejecting contact request
  React.useEffect(() => {
    const handleRejectPendingContacts = (data) => {
      setUserPending((prevPending) => {
        const updatedPending = prevPending.filter(
          (pending) => pending._id !== data._id,
        );
        return updatedPending;
      });
    };
    socket.on("RejectContactRequest", handleRejectPendingContacts);
    return () => {
      socket.off("RejectContactRequest", handleRejectPendingContacts);
    };
  }, [userPending]);

  // Filter contacts based on selected tab
  React.useEffect(() => {
    if (!userContacts) return;

    const onlineContacts = userContacts.filter(
      (contact) => contact.status === "Online",
    );

    dispatch(setOnlineAmount(onlineContacts.length));
    dispatch(setContactsAmount(userContacts.length));
    dispatch(setPendingAmount(userPending.length));

    // Sort contacts by username and alphabetically
    userContacts.sort((a, b) => a.username.localeCompare(b.username));

    switch (selectedTab) {
      // Online tab selected
      case 0:
        setFilteredContacts(onlineContacts);
        break;
      // All tab selected
      case 1:
        setFilteredContacts(userContacts);
        break;
      // Pending tab selected
      case 2:
        setFilteredContacts(sortPendingContacts(userPending));
        break;
      default:
        setFilteredContacts([]);
    }
  }, [userContacts, userPending, selectedTab]);

  React.useEffect(() => {
    if (!searchValue && filteredContacts) {
      setSearchedContacts(filteredContacts);
      return;
    }
    if (searchValue) {
      const displayedsearchedContacts = filteredContacts.filter((contact) =>
        contact.username.toLowerCase().includes(searchValue.toLowerCase()),
      );
      setSearchedContacts(displayedsearchedContacts);
    }
  }, [searchValue, filteredContacts]);

  const handleSearchChange = (e) => {
    setSearchValue(e.target.value);
  };

  return (
    <Container data-testid="contact-page" maxWidth="md">
      <Toolbar />
      <ContactTabs selectedTab={selectedTab} setSelectedTab={setSelectedTab} />
      <Stack direction="column" spacing={2}>
        <SearchBar
          searchValue={searchValue}
          handleSearchChange={handleSearchChange}
        />
        <Divider />
        <Stack
          direction="column"
          spacing={0}
          sx={{
            maxHeight: "calc(100vh - 210px)",
            overflowY: "scroll",
          }}
        >
          {searchedContacts.map((contact) => (
            // eslint-disable-next-line no-underscore-dangle
            <ContactCard key={contact._id} contact={contact} />
          ))}
        </Stack>
      </Stack>
    </Container>
  );
}

export default Contacts;
