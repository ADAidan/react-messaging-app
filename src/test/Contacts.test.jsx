import * as React from "react";
import { render, screen } from "@testing-library/react";
import Contacts from "../pages/Contacts";

test("renders Contacts page", () => {
  render(<Contacts />);
  const contactPage = screen.getByTestId("contact-page");
  expect(contactPage).toBeInTheDocument();
});
