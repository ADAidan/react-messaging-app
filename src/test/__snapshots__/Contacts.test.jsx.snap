// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Contacts > should render the Contacts page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
        data-testid="contact-page"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
        />
        <div
          class="MuiBox-root css-1cip7rb"
        >
          <div
            class="MuiStack-root css-3y1137-MuiStack-root"
          >
            <div
              class="MuiTabs-root css-1ujnqem-MuiTabs-root"
            >
              <div
                class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-oqr85h"
                style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
              />
              <div
                class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
                style="margin-bottom: 0px;"
              >
                <div
                  aria-label="contact tabs"
                  class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                  role="tablist"
                >
                  <button
                    aria-label="Online contacts"
                    aria-selected="true"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    Online 0
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-label="All contacts"
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    All 0
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-label="Pending contacts"
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    Pending 0
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <span
                  class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-ke5b6m-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Add Contact
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiStack-root css-1sazv7p-MuiStack-root"
        >
          <div
            class="MuiBox-root css-pppzjk"
            data-testid="search-bar"
          >
            <div
              class="MuiBox-root css-1i1gqge"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary css-1776fzg-MuiInputBase-root"
            >
              <input
                aria-label="search"
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                placeholder="Search…"
                type="text"
                value=""
              />
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
          <div
            class="MuiStack-root css-1kdmtr3-MuiStack-root"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
      data-testid="contact-page"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiBox-root css-1cip7rb"
      >
        <div
          class="MuiStack-root css-3y1137-MuiStack-root"
        >
          <div
            class="MuiTabs-root css-1ujnqem-MuiTabs-root"
          >
            <div
              class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-oqr85h"
              style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
            />
            <div
              class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
              style="margin-bottom: 0px;"
            >
              <div
                aria-label="contact tabs"
                class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  aria-label="Online contacts"
                  aria-selected="true"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  Online 0
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="All contacts"
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  All 0
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Pending contacts"
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  Pending 0
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <span
                class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess MuiButton-root MuiButton-contained MuiButton-containedSuccess MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSuccess css-ke5b6m-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Add Contact
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-pppzjk"
          data-testid="search-bar"
        >
          <div
            class="MuiBox-root css-1i1gqge"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
          </div>
          <div
            class="MuiInputBase-root MuiInputBase-colorPrimary css-1776fzg-MuiInputBase-root"
          >
            <input
              aria-label="search"
              class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
              placeholder="Search…"
              type="text"
              value=""
            />
          </div>
        </div>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        />
        <div
          class="MuiStack-root css-1kdmtr3-MuiStack-root"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

